hail/src/main/scala/is/hail/cxx/PackDecoder.scala

```diff
@@ -0,0 +1,240 @@
+      }.mkString("\n")
+      if (t.nMissingBytes > 0) {
+        s"""
+           |$input_buf_ptr->read_bytes($off, ${ t.nMissingBytes });
+           |$decodeFields""".stripMargin
+      } else
+        decodeFields
+    } else {
+      val names = t.asInstanceOf[PStruct].fieldNames
+      val rnames = rt.asInstanceOf[PStruct].fieldNames
+      val t_missing_bytes = ArrayVariable(s"missing", "char", s"${ t.nMissingBytes }")
+
+      var rtidx = 0
+      val decodeFields = Array.tabulate[Code](t.size) { tidx =>
+        val f = t.types(tidx)
+        val skipField = rtidx >= rt.size || names(tidx) != rnames(rtidx)
+        val processField = if (skipField)
+          wrapMissing(s"$t_missing_bytes", tidx)(skip(f, input_buf_ptr))
+        else
+          s"""
+             |${ wrapMissing(s"$t_missing_bytes", tidx)(ssb.addField(rtidx, foff => decode(f, rt.types(rtidx), input_buf_ptr, region, Expression(foff), fb))) }
+             |${ if (f.required) "" else s" else { ${ ssb.setMissing(rtidx) } }" }
+           """.stripMargin
+        if (!skipField)
+          rtidx += 1
+        wrapMissing(s"$t_missing_bytes", tidx)(processField)
+      }.mkString("\n")
+      if (t.nMissingBytes > 0) {
+        s"""
+           |${ t_missing_bytes.define }
+           |${ ssb.clearAllMissing() };
+           |$input_buf_ptr->read_bytes($t_missing_bytes, ${ t.nMissingBytes });
+           |$decodeFields""".stripMargin
+      } else
+        decodeFields
+    }
+  }
+
+  def decode(t: PType, rt: PType, input_buf_ptr: Expression, region: Expression, off: Expression, fb: FunctionBuilder): Code = t match {
+    case _: PBoolean => s"store_byte($off, $input_buf_ptr->read_byte());"
+    case _: PInt32 => s"store_int($off, $input_buf_ptr->read_int());"
+    case _: PInt64 => s"store_long($off, $input_buf_ptr->read_long());"
+    case _: PFloat32 => s"store_float($off, $input_buf_ptr->read_float());"
+    case _: PFloat64 => s"store_double($off, $input_buf_ptr->read_double());"
+    case _: PBinary => decodeBinary(input_buf_ptr, region, off, fb)
+    case t2: PArray => decodeArray(t2, rt.asInstanceOf[PArray], input_buf_ptr, region, off, fb)
+    case t2: PBaseStruct => decodeBaseStruct(t2, rt.asInstanceOf[PBaseStruct], input_buf_ptr, region, off, fb)
+  }
+
+  def apply(t: PType, rt: PType, bufSpec: BufferSpec, tub: TranslationUnitBuilder): Class = {
+    tub.include("hail/hail.h")
+    tub.include("hail/Decoder.h")
+    tub.include("hail/Region.h")
+    tub.include("hail/Utils.h")
+    tub.include("<cstdio>")
+    tub.include("<memory>")
+
+    val decoderBuilder = new ClassBuilder(genSym("Decoder"), "public NativeObj")
+
+    val bufType = bufSpec.nativeInputBufferType
+    val buf = Variable("buf", s"std::shared_ptr<$bufType>")
+    decoderBuilder.addPrivate(buf)
+
+    decoderBuilder.addConstructor(s"${ decoderBuilder.name }(std::shared_ptr<InputStream> is) { $buf = std::make_shared<$bufType>(is); }")
```

Comment:
I think
```
s"${ decoderBuilder.name }(std::shared_ptr<InputStream> is) : $buf(std::make_shared<$bufType>(is)) {}"
```
works here, and you should always initialize all variables in the member initializer list when possible.